src/Protocol/ForgeHandshake.cpp

```diff
@@ -0,0 +1,352 @@
+enum
+{
+	Discriminator_ServerHello = 0,
+	Discriminator_ClientHello = 1,
+	Discriminator_ModList = 2,
+	Discriminator_RegistryData = 3,
+	Discriminator_HandshakeReset = -2,
+	Discriminator_HandshakeAck = -1,
+};
+
+/** Client handshake state phases. */
+enum
+{
+	ClientPhase_WAITINGSERVERDATA = 2,
+	ClientPhase_WAITINGSERVERCOMPLETE = 3,
+	ClientPhase_PENDINGCOMPLETE = 4,
+	ClientPhase_COMPLETE = 5,
+};
+
+/** Server handshake state phases. */
+enum
+{
+	ServerPhase_WAITINGCACK = 2,
+	ServerPhase_COMPLETE = 3,
+};
+
+
+
+
+
+cForgeHandshake::cForgeHandshake(cClientHandle *client) :
+	m_IsForgeClient(false),
+	m_Errored(false),
+	m_Client(client)
+{
+}
+
+
+
+
+
+void cForgeHandshake::AugmentServerListPing(Json::Value & a_ResponseValue)
+{
+	UInt32 ProtocolVersion = m_Client->GetProtocolVersion();
+	cForgeMods & Mods = cRoot::Get()->GetServer()->GetRegisteredForgeMods(ProtocolVersion);
+
+	if (!Mods.HasMods())
+	{
+		return;
+	}
+
+	LOG("Received server ping from version: %d", ProtocolVersion);
+
+	// modinfo:
+	Json::Value Modinfo;
+	Modinfo["type"] = "FML";
+
+	Json::Value ModList(Json::arrayValue);
+
+	for (auto & item: Mods.GetMods())
+	{
+		Json::Value Mod;
+		Mod["modid"] = item.first;
+		Mod["version"] = item.second;
+		ModList.append(Mod);
+	}
+
+	Modinfo["modList"] = ModList;
+
+	// Augment the response:
+	a_ResponseValue["modinfo"] = Modinfo;
+}
+
+
+
+
+
+void cForgeHandshake::BeginForgeHandshake(const AString & a_Name, const AString & a_UUID, const Json::Value & a_Properties)
+{
+	ASSERT(m_IsForgeClient);
+
+	m_Name = &a_Name;
+	m_UUID = &a_UUID;
+	m_Properties = &a_Properties;
+
+	AStringVector Channels = { "FML|HS", "FML", "FML|MP", "FML", "FORGE" };
+	AString ChannelsString;
+
+	for (AStringVector::iterator itr = Channels.begin(); itr != Channels.end(); ++itr)
+	{
+		ChannelsString.append(*itr);
+		ChannelsString.push_back('\0');
+	}
+
+	m_Client->SendPluginMessage("REGISTER", ChannelsString);
+	SendServerHello();
+}
+
+
+
+
+
+void cForgeHandshake::SendServerHello()
+{
+	AString Message;
+	// Discriminator | Byte | Always 0 for ServerHello
+	Message.push_back(Discriminator_ServerHello);
+	// FML protocol Version | Byte | Determined from NetworkRegistery. Currently 2.
+	Message.push_back('\2');
+	// Dimension TODO
+	Message.push_back('\0');
+	Message.push_back('\0');
+	Message.push_back('\0');
+	Message.push_back('\0');
+
+	m_Client->SendPluginMessage("FML|HS", Message);
+}
+
+
+
+
+
+AStringMap cForgeHandshake::ParseModList(const char * a_Data, size_t a_Size)
+{
+	AStringMap Mods;
+
+	if (a_Size < 4)
+	{
+		LOG("ParseModList invalid packet, missing length (size=%zu)", a_Size);
+		return Mods;
+	}
+
+	cByteBuffer Buf(a_Size);
+	Buf.Write(a_Data, a_Size);
+
+	UInt32 NumMods;
+	Buf.ReadVarInt32(NumMods);
+
+	for (size_t i = 0; i < NumMods; ++i)
+	{
+		AString Name, Version;
+		Buf.ReadVarUTF8String(Name);
+		Buf.ReadVarUTF8String(Version);
+
+		Mods.insert(std::pair<AString, AString>(Name, Version));
```

Comment:
`insert` here actually takes a `std::pair<const AString, AString>` but it's probably cleanest to just use uniform initialisation:
```
Mods.insert({ Name, Version });
```